gcsstore/gcsstore.go

```diff
@@ -0,0 +1,224 @@
+package gcsstore
+
+import (
+	"io"
+	"bytes"
+	"fmt"
+	"strings"
+	"strconv"
+	"encoding/json"
+
+	"github.com/tus/tusd"
+	"github.com/tus/tusd/uid"
+)
+
+type GCSStore struct {
+	Bucket string
+	Service GCSAPI
+}
+
+// ~/go/bin/mockgen -destination ~/go/src/github.com/tus/tusd/gcsstore/gcsstore_mock_test.go -package=gcsstore_test github.com/tus/tusd/gcsstore GCSReader,GCSAPI
+
+func New(bucket string, service GCSAPI) GCSStore {
+	return GCSStore {
+		Bucket: bucket,
+		Service: service,
+	}
+}
+
+func (store GCSStore) UseIn(composer *tusd.StoreComposer) {
+	composer.UseCore(store)
+	composer.UseTerminater(store)
+	composer.UseFinisher(store)
+	composer.UseGetReader(store)
+}
+
+func (store GCSStore) NewUpload(info tusd.FileInfo) (id string, err error) {
+	if info.ID == "" {
+		info.ID = uid.Uid()
+	}
+
+	err = store.WriteInfo(info.ID, info)
+	if err != nil {
+		return info.ID, err
+	}
+
+	return info.ID, nil
+}
+
+func (store GCSStore) WriteChunk(id string, offset int64, src io.Reader) (int64, error) {
+	info, err := store.GetInfo(id)
+	if err != nil {
+		return 0, err
+	}
+
+	prefix := fmt.Sprintf("%s_", id)
+	filterParams := GCSFilterParams {
+		Bucket: store.Bucket,
+		Prefix: prefix,
+	}
+
+	names, err := store.Service.FilterObjects(filterParams)
+	maxIdx := -1
+
+	for _, name := range names {
+		split := strings.Split(name, "_")
+		idx, err := strconv.Atoi(split[len(split)-1])
+		if err != nil {
+			return -1, err
+		}
+
+		if idx > maxIdx {
+			maxIdx = idx
+		}
+	}
+
+	cid := fmt.Sprintf("%s_%d", id, maxIdx + 1)
+	objectParams := GCSObjectParams {
+		Bucket: store.Bucket,
+		ID: cid,
+	}
+
+	n, err := store.Service.WriteObject(objectParams, src)
+	if err != nil {
+		return 0, err
+	}
+
+	info.Offset = info.Offset + n
+	err = store.WriteInfo(id, info)
```

Comment:
If the call to WriteInfo fails (for whatever reason), you can end up in an inconsistent state where you were able to upload the chunk to the bucket but didn't update the offset. In order to avoid this issue, the s3store does not update the offset after writing a chunk but instead calculates the offset for each call to GetInfo (see https://github.com/tus/tusd/blob/master/s3store/s3store.go#L309-L335). It would be nice to have a similar approach here.